Fix blend shape handling.

bqqbarbhg opened the pull request:
Should fix #13.

The core of the issue here is that geometry data in FBX is not indexed, so there is no way to map from the processed array mesh into blend shape offsets after `mesh_surface_tool->index()` is called. I assume that blend shapes are currently broken for non-indexed glTF meshes as well.

To fix this we need to retain vertex index information during this process, but as we have no extra geometry data channels to use currently I did a somewhat strange fix: Instead of storing vertex positions directly in `Mesh::ARRAY_VERTEX`, for meshes with blend shapes we store instead `Vector3(index & 0xffff, index >> 16, 0)`. This prevents the `index()` call from merging vertices that are in the same position but have potentially different blend offsets, but still allows merging identical vertices by index.

Afterwards it's easy to retrieve the vertex indices for generating the blend shape data. The real vertex positions are later patched in after blend shapes are read.

Edited:

Fixes: https://github.com/V-Sekai/godot-ufbx/issues/10
Fixes: https://github.com/V-Sekai/godot-ufbx/issues/13

bqqbarbhg commented:
I also added a temporary local modification to ufbx to support normals in blend shapes, I'll add it to mainline ufbx with some tests soon.

fire commented:
The normals are broken. Maybe axis conversion?

https://github.com/V-Sekai/godot-ufbx/assets/32321/5a187b4c-6a48-4887-89a9-a02816a46d56

bqqbarbhg commented:
Shouldn't be it's local space, I'll take a look with that model!

bqqbarbhg commented:
Turns out the normals were offsets as well (which is not too surprising), they feel a bit too strong in this model so not 100% sure what's going on but they look roughly correct at least.

fire commented:
`array[Mesh::ARRAY_NORMAL] = _decode_vertex_attrib_vec3(fbx_mesh->vertex_tangent, indices);`

Noticed this bug.

bqqbarbhg commented:
Whelp how did that happen, awkward copy-paste mistake from me, I'll fix it in this PR and see if it helps.

fire commented:
Forgot to mention Tangent is a vector4/plane/color.

See the other uses of ARRAY_TANGENT

bqqbarbhg commented:
Oh it contains the bitangent sign as well?

fire commented:
See https://godotengine.org/article/fixing-tangents-accurate-normal-maps/

fire commented:
> To save on space Godot uses the approach where only the tangents are store and bi-tangents are calculated by combining them with the normal vector. For this a direction flag is also included with each tangent.

bqqbarbhg commented:
Yeah that's what I expected, I'll fix it!

fire commented on fbx_document.cpp:
```diff
@@ -1791,31 +1818,49 @@ Error FBXDocument::_parse_meshes(Ref<FBXState> p_state) {
+								Vector3 *w_varr = varr.ptrw();
+								Vector3 *w_narr = narr.ptrw();
+								const Vector3 *r_varr = src_varr.ptr();
+								const Vector3 *r_narr = src_narr.ptr();
```
```suggestion
```
This variable is unused?

bqqbarbhg replied:
Whoops, I read straight from `src_narr` instead. I think this is just an optimization in the original source that I kept.

fire commented on thirdparty/ufbx/ufbx.c:
```diff
@@ -26524,18 +26524,25 @@ ufbx_abi ufbxi_noinline ufbx_matrix ufbx_catch_get_skin_vertex_matrix(ufbx_panic
 {
 	ufbx_assert(shape);
-	if (!shape) return ufbx_zero_vec3;
+	if (!shape) return UFBX_NO_INDEX;
```
Code style.

We always use curly braces.

fire replied:
Oops this is in ufbx.

bqqbarbhg committed:
Fix blend shape handling.

ufbx: Add ufbx_get_blend_shape_offset_index()

Fix blend shape normal vector handling.

Fix tangent array destination.

Fix meshes without blend shapes.

Fix imported tangent format.

Use r_narr for performance.

fire commented:
When we're ready, we can rebase to one commit and merge. The reasoning for one commit is we ran into a detour in the middle of the patch series.

bqqbarbhg commented:
Sure, do you mind if I add temporary animation support here as well while I'm at it?

fire commented:
Want me to merge this in first? 

I can do it promptly. 

Then you can start on temporary animation support.

bqqbarbhg commented:
Sure thing! I'm still sorting some unrelated stuff out for a while before I could start.
